src/library_manager/lib_manager.c

```diff
@@ -68,114 +68,133 @@ static int lib_manager_load_data_from_storage(void __sparse_cache *vma, void *s_
 	return 0;
 }
 
-static int lib_manager_load_module(uint32_t module_id, struct sof_man_module *mod,
-				   struct sof_man_fw_desc *desc)
+static int lib_manager_load_module(const uint32_t module_id,
+				   const struct sof_man_module *const mod)
 {
-	struct ext_library *ext_lib = ext_lib_get();
-	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
-	size_t load_offset = (size_t)((void *)ext_lib->desc[lib_id]);
-	void __sparse_cache *va_base_text = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
-	void *src_txt = (void *)(mod->segment[SOF_MAN_SEGMENT_TEXT].file_offset + load_offset);
-	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
-	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
-	void *src_rodata =
-		(void *)(mod->segment[SOF_MAN_SEGMENT_RODATA].file_offset + load_offset);
-	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
+	const struct ext_library *const ext_lib = ext_lib_get();
+	const uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	const uintptr_t load_offset = POINTER_TO_UINT(ext_lib->desc[lib_id]);
+	void *src, *va_base;
+	size_t size;
+	uint32_t flags;
+	int ret, idx;
+
+	for (idx = 0; idx < ARRAY_SIZE(mod->segment); ++idx) {
```

Comment:
on the one hand it's good to be flexible, to parse the input data safely, but OTOH aren't segment array index values a part of the ABI? Can `SOF_MAN_SEGMENT_*` values really be changed freely?

Reply:
@lyakh  I checked the ext rom code and it looks like it uses static segment indexes.